### ISSUE TITLE
2.4: "Balance due" payment state after reimbursing a return

### ISSUE THREAD

**Issue by frahugo** [spree/spree#6229]
Here's how to reproduce using the sandbox app:

1. Place an order on the front-end
2. Ship the shipment
3. Create a Return Authorization (select the Original Reimbursement type)
4. Create a Customer Return (select the item from 'Items in Return Authorizations')
5. Create reimbursement
6. Reimburse.

The payment state goes from "Paid" to "Balance due".  I guess it should stay to "Paid".

**Comment by JDutil** [comment]
Would this be a duplicate of https://github.com/spree/spree/issues/6226 caused by a rounding issue?  Or are the payment totals even more off than a cent or two?

**Comment by frahugo** [comment]
I don't think it's a rounding issue.  The order has only one item in it: Ruby on Rails Tote @ 15.99.  Plus 5.00 for the shipping, and 0.80 for tax.  Total of 21.79.  The return authorization was created for the tote, for the amount of 15.99.  So the return and reimbursement are also 15.99.

Through the console, order#outstanding_balance gives me 16.79.  From what I understand, the payment_state is based on the outstanding balance amount.  So this is clearly not a rounding issue.

**Comment by JDutil** [comment]
Have you tried updating to latest 2-4-stable? What's your Gemfile.lock I'm pretty sure we had already fixed this bug.

**Comment by frahugo** [comment]
Let me fetch the latest 2-4 and recreate the sandbox.

**Comment by frahugo** [comment]
Same behaviour with latest 2-4-stable branch (last commit: 19307ae7).  I made a recording: http://hf-monosnap.s3.amazonaws.com/screencast_2015-03-26_13-58-21_0_edited.mp4

**Comment by JDutil** [comment]
I've confirmed this is an issue on master still as well. I think the issue was caused by a change I made for the related issue I was thinking of here:
https://github.com/spree/spree/issues/5907

Essentially we were seeing that if you created an order adjustment that for instance would give $10 off the order total then issued a $10 refund the payment state would remain as credit owed. With the fix to that being now to calculate the payment total to include the refunds we're now inadvertently causing this new issue:
https://github.com/spree/spree/commit/10816268c0e48515a1951f11c841b3a835cd00e5

I'm not sure what the best solution to this will be that the community would also like, but I'm thinking we could create a line item adjustment for returned items in order to lower the order total. The drawback I see to this though is that it messes with the orders summary and potentially peoples sales reporting depending on how they do it. Which leads me to believe attempting to fix this issue may also inadvertently cause other issues that may or may not be worse than the payment state being "balance due".

The other thing I'll try doing is changing the Order#outstanding_balance method to also consider order returns, but that will have to further degrade performance with even more database calls during the OrderUpdater.
